pxr/imaging/plugin/hdEmbree/ies.cpp

```diff
@@ -0,0 +1,490 @@
+        *(data++) = bit_cast<int, float>(_v_angles.size());
+
+        memcpy(data, &_h_angles[0], _h_angles.size() * sizeof(float));
+        data += _h_angles.size();
+        memcpy(data, &_v_angles[0], _v_angles.size() * sizeof(float));
+        data += _v_angles.size();
+
+        for (int h = 0; h < _intensities.size(); h++) {
+            memcpy(data, &_intensities[h][0], _v_angles.size() * sizeof(float));
+            data += _v_angles.size();
+        }
+    }
+}
+
+class IESTextParser {
+public:
+    std::vector<char> text;
+    char *data;
+    bool error;
+
+    IESTextParser(const std::string &str)
+        : text(str.begin(), str.end()), error(false) {
+        std::replace(text.begin(), text.end(), ',', ' ');
+        data = strstr(&text[0], "\nTILT=");
+    }
+
+    bool eof() { return (data == NULL) || (data[0] == '\0'); }
+
+    bool has_error() { return error; }
+
+    double get_double() {
+        if (eof()) {
+            error = true;
+            return 0.0;
+        }
+        char *old_data = data;
+        double val = strtod(data, &data);
+        if (data == old_data) {
+            data = NULL;
+            error = true;
+            return 0.0;
+        }
+        return val;
+    }
+
+    long get_long() {
+        if (eof()) {
+            error = true;
+            return 0;
+        }
+        char *old_data = data;
+        long val = strtol(data, &data, 10);
+        if (data == old_data) {
+            data = NULL;
+            error = true;
+            return 0;
+        }
+        return val;
+    }
+};
+
+bool IESFile::parse(const std::string &ies) {
+    if (ies.empty()) {
+        return false;
+    }
+
+    IESTextParser parser(ies);
+    if (parser.eof()) {
+        return false;
+    }
+
+    // Handle the tilt data block
+    if (strncmp(parser.data, "\nTILT=INCLUDE", 13) == 0) {
+        parser.data += 13;
+        parser.get_double();              //< Lamp to Luminaire geometry
+        int num_tilt = parser.get_long(); //< Amount of tilt angles and factors
+
+        // Skip over angles and factors
+        for (int i = 0; i < 2 * num_tilt; i++) {
+            parser.get_double();
+        }
+    } else {
+        // Skip to next line
+        parser.data = strstr(parser.data + 1, "\n");
+    }
+
+    if (parser.eof()) {
+        return false;
+    }
+    parser.data++;
+
+    parser.get_long();                    //< Number of lamps
+    parser.get_double();                  //< Lumens per lamp
+    double factor = parser.get_double();  //< Candela multiplier
+    int v_angles_num = parser.get_long(); //< Number of vertical angles
+    int h_angles_num = parser.get_long(); //< Number of horizontal angles 
+    _type = (IESType)parser.get_long();   //< Photometric type
+
+    // TODO(lukas): Test whether the current type B processing can also deal with
+    // type A files. In theory the only difference should be orientation which we
+    // ignore anyways, but with IES you never know...
```

Comment:
see my note on ies.h. The formula I posted is from IESNA LM-75-01, so we don't have to guess :)